File: frontend/src/types/Deck.ts
import { DeckCategory } from "./DeckCategory";
import { ImageFile } from "./ImageFile";

export interface Deck {
  id: string;
  userId: string;
  username: string;
  title: string;
  description: string;
  category: string;
  tags: string[];
  isPublic: boolean;
  imageUrl?: string;
  flashcardsCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface DeckWithRatingResponse extends Deck {
  rating: number;
  totalVotes: number;
  upvoteCount: number;
}

export interface CreateDeckRequest {
  userId: string;
  title: string;
  description: string;
  category: string;
  tags: string[];
  isPublic: boolean;
  imageUrl?: string | null;
}

export interface UpdateDeckRequest {
  title: string;
  description: string;
  category: string;
  tags: string[];
  isPublic: boolean;
  imageUrl?: string | null;
}

export interface DeckFormData {
  title: string;
  description: string;
  category: DeckCategory;
  tags?: string;
  isPublic?: boolean;
  imageFile?: ImageFile | null;
  consent: boolean;
}

export const isDeckWithRating = (
  deck: Deck | DeckWithRatingResponse
): deck is DeckWithRatingResponse => {
  return "rating" in deck && "totalVotes" in deck && "upvoteCount" in deck;
};

export const convertFormToRequest = (
  formData: DeckFormData,
  userId: string
): CreateDeckRequest => ({
  userId,
  title: formData.title,
  description: formData.description,
  category: formData.category,
  tags:
    formData.tags
      ?.split(",")
      .map((tag) => tag.trim())
      .filter(Boolean) ?? [],
  isPublic: formData.isPublic ?? true,
});

File: backend/Lithuaningo.API/Models/Deck.cs
using System;
using Supabase.Postgrest.Models;
using Supabase.Postgrest.Attributes;

namespace Lithuaningo.API.Models
{
    [Table("decks")]
    public class Deck : BaseModel
    {
        [PrimaryKey("id")]
        [Column("id")]
        public Guid Id { get; set; }

        [Column("user_id")]
        public Guid UserId { get; set; }

        [Column("user_name")]
        public string UserName { get; set; } = string.Empty;

        [Column("title")]
        public string Title { get; set; } = string.Empty;

        [Column("description")]
        public string Description { get; set; } = string.Empty;

        [Column("category")]
        public string Category { get; set; } = string.Empty;

        [Column("tags")]
        public string[] Tags { get; set; } = Array.Empty<string>();

        [Column("flashcard_count")]
        public int FlashcardCount { get; set; }

        [Column("rating")]
        public double Rating { get; set; }

        [Column("image_url")]
        public string? ImageUrl { get; set; }

        [Column("is_public")]
        public bool IsPublic { get; set; }

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        [Column("updated_at")]
        public DateTime UpdatedAt { get; set; }
    }
}
